neuraxle/metaopt/trial.py

```diff
@@ -50,32 +55,224 @@ def __init__(
+            'end_time': self.end_time.strftime(TRIAL_DATETIME_STR_FORMAT) if self.end_time is not None else ''
+        }
+
+    @staticmethod
+    def from_json(trial_json) -> 'Trial':
+        return Trial(
+            status=TRIAL_STATUS(trial_json['status']),
+            hyperparams=HyperparameterSamples(trial_json['hyperparams']),
+            validation_splits=[
+                TrialSplit.from_json(validation_split_json)
+                for validation_split_json in trial_json['validation_splits']
+            ],
+            error=trial_json['error'],
+            error_traceback=trial_json['error_traceback'],
+            start_time=datetime.datetime.strptime(trial_json['start_time'], TRIAL_DATETIME_STR_FORMAT),
+            end_time=datetime.datetime.strptime(trial_json['start_time'], TRIAL_DATETIME_STR_FORMAT)
+        )
+
+    def __enter__(self):
+        """
+        Start trial, and set the trial status to PLANNED.
+        """
+        self.start_time = datetime.datetime.now()
+        self.status = TRIAL_STATUS.STARTED
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        """
+        Stop trial, and save end time.
+
+        :param exc_type:
+        :param exc_val:
+        :param exc_tb:
+        :return:
+        """
+        self.end_time = datetime.datetime.now()
+        del self.pipeline
+        return self
+
+
+class TrialSplit:
+    """
+    Trial split
+
+    .. seealso::
+        :class:`AutoML`,
+        :class:`HyperparamsRepository`,
+        :class:`BaseHyperparameterSelectionStrategy`,
+        :class:`RandomSearchHyperparameterSelectionStrategy`,
+        :class:`DataContainer`
+    """
+
+    def __init__(
+            self,
+            split_number,
+            status: 'TRIAL_STATUS' = None,
             error: Exception = None,
             error_traceback: str = None,
             metrics_results: Dict = None,
-            pipeline: BaseStep = None,
-            cache_folder = None
+            start_time: datetime.datetime = None,
+            end_time: datetime.datetime = None,
+            pipeline: BaseStep = None
     ):
         if status is None:
             status = TRIAL_STATUS.PLANNED
-        self.cache_folder = cache_folder
+        self.split_number = split_number
         self.status: TRIAL_STATUS = status
         self.error: Exception = error
         self.error_traceback: str = error_traceback
         if metrics_results is None:
             metrics_results = {}
         self.metrics_results: Dict = metrics_results
-        self.pipeline: BaseStep = pipeline
-        self.hyperparams: HyperparameterSamples = hyperparams
+        self.end_time = end_time
+        self.start_time = start_time
+        self.pipeline = pipeline
 
-    def set_fitted_pipeline(self, pipeline: BaseStep):
-        """
-        Set fitted pipeline.
+    def fit(self, train_data_container: DataContainer, context: ExecutionContext) -> 'TrialSplit':
```

Comment:
This is not a base step. I'd avoid reusing the same name "fit" as it's misleading. Maybe fit_pipeline ?

(same comment as for different versions of the methods `save` and `hash` that we had in the past)